File: 131. Palindrome Partitioning.js
/**
 * @param {string} s
 * @return {string[][]}
 */
var partition = function(s) {
    var result = [];
    if ( s == null || s.length == 0 ) {
        return result;
    }

    dfs(s, 0, result, []);

    return result;

    function dfs( str, startIndex, result, tmp ){
        if ( startIndex == str.length ) {
            result.push(tmp.slice());
        }

        for ( var i = startIndex; i < str.length; i++ ) {
            var subString = str.slice(startIndex, i + 1);
            if ( !isPalindrome(subString) ) {
                continue;
            }
            tmp.push(subString);
            dfs(str, i + 1, result, tmp);
            tmp.pop();

        }
    }

    function isPalindrome ( str ) {
        if ( str == null || str.length == 0 ) {
            return false;
        }

        for ( var i = 0; i < str.length; i++ ) {
            if ( str[i] != str[str.length - 1 - i] ) {
                return false;
            }

            if ( i == str.length - i ) {
                break;
            }
        }

        return true;
    }
};

File: 131. Palindrome Partitioning.py
"""
Given a string s, 
partition s such that every substring of the partition is a palindrome.
Return all possible palindrome partitioning of s.

Example:
Input: "aab"
Output:
[
  ["aa","b"],
  ["a","a","b"]
]
"""
class Solution(object):
    def partition(self, s):
        self.result = []
        self.dfs(s, [])
        return self.result

    def dfs(self, s, path):
        if len(s) == 0:
            self.result.append(path)
            return
        for i in xrange(1, len(s) + 1):
            word = s[:i]
            if self.isPalindrome(word):
                self.dfs(s[i:], path + [s[:i]])

    def isPalindrome(self, s):
        start = 0
        end = len(s) - 1
        while start < end:
            if s[start] != s[end]:
                return False
            start += 1
            end -= 1
        return True
